### PR TITLE
Add keyboard shortcuts and check highlights

### PR DESCRIPTION
**Pull request by lln-ijinus**
A proposition to add new shortcuts :
- Alt+v to copy source to target
- Esc then a number to remove a check
- Alt then a number to copy a machine translation
- Alt+(Char) to copy an highlighted check from source to cursor position

All these modifications are done to help translator to be efficient by not having to use the mouse.

To be able to add the last shortcuts, <span> are created around parts of the source string that matches a check. For now around source parts that checks formats( %s, {0}, etc.) and markup.

To do that, a new method "check_highlight" was added to the base check.

Most of the work is in the javascript to handle shortcuts when they are needed and have a little impact on the code.
For the shortcuts to remove checks, to copy a machine translation and to copy highlight checks where the shortcut is "variable", infos are given on the number on the given elements (<sup>1</sup>, etc.) and these informations are also added only in the javascript.

### CODE REVIEW

File: weblate/trans/checks/markup.py [hunk 1]
```diff
@@ -64,6 +64,13 @@ def check_single(self, source, target, unit):
 
         return src_tags != tgt_tags
 
+    def check_highlight(self,source,unit):
+        ret=[]
+        matchObjects = BBCODE_MATCH.finditer(source)
+        for m in matchObjects:
+            if m.start()==m.end(): continue
```
Comment (nijel) [inline weblate/trans/checks/markup.py]: Codacy Issue found: [More than one statement on a single line](https://www.codacy.com/app/nijel/weblate/file/1546310447/issues/source?bid=2038971&fileBranchId=3014792#l71)

Comment (nijel) [inline weblate/trans/checks/markup.py]: Codacy Issue found: [Exactly one space required around comparison](https://www.codacy.com/app/nijel/weblate/file/1546310447/issues/source?bid=2038971&fileBranchId=3014792#l71)

File: weblate/trans/checks/markup.py [hunk 2]
```diff
@@ -64,6 +64,13 @@ def check_single(self, source, target, unit):
...
+            ret.append((m.start(),m.group()))
```
Comment (nijel) [inline weblate/trans/checks/markup.py]: Codacy Issue found: [Exactly one space required after comma](https://www.codacy.com/app/nijel/weblate/file/1546310447/issues/source?bid=2038971&fileBranchId=3014792#l72)

File: weblate/trans/checks/markup.py [hunk 3]
```diff
@@ -64,6 +64,13 @@ def check_single(self, source, target, unit):
 
         return src_tags != tgt_tags
 
+    def check_highlight(self,source,unit):
+        ret=[]
```
Comment (nijel) [inline weblate/trans/checks/markup.py]: Codacy Issue found: [Exactly one space required around assignment](https://www.codacy.com/app/nijel/weblate/file/1546310447/issues/source?bid=2038971&fileBranchId=3014792#l68)

File: weblate/trans/checks/markup.py [hunk 4]
```diff
@@ -64,6 +64,13 @@ def check_single(self, source, target, unit):
 
         return src_tags != tgt_tags
 
+    def check_highlight(self,source,unit):
```
Comment (nijel) [inline weblate/trans/checks/markup.py]: Codacy Issue found: [Exactly one space required after comma](https://www.codacy.com/app/nijel/weblate/file/1546310447/issues/source?bid=2038971&fileBranchId=3014792#l67)

File: weblate/trans/checks/markup.py [hunk 5]
```diff
@@ -103,3 +110,15 @@ def check_single(self, source, target, unit):
...
+        ret=[]
+        matchObjects = XML_MATCH.finditer(source)
+        for m in matchObjects:
+            if m.start()==m.end(): continue
+            ret.append((m.start(),m.group()))
+        matchObjects = XML_ENTITY_MATCH.finditer(source)
+        for m in matchObjects:
+            if m.start()==m.end(): continue
+            ret.append((m.start(),m.group()))
+        return ret
```
Comment (nijel) [inline weblate/trans/checks/markup.py]: Codacy Issue found: [Final newline missing](https://www.codacy.com/app/nijel/weblate/file/1546310447/issues/source?bid=2038971&fileBranchId=3014792#l124)

File: weblate/trans/checks/format.py [hunk 6]
```diff
@@ -203,6 +203,13 @@ def check_single(self, source, target, unit):
         '''
         return False
 
+    def check_highlight(self,source,unit):
+        ret=[]
+        matchObjects = self.regexp.finditer(source)
+        for m in matchObjects:
+            if m.start()==m.end(): continue
```
Comment (nijel) [inline weblate/trans/checks/format.py]: Codacy Issue found: [Exactly one space required around comparison](https://www.codacy.com/app/nijel/weblate/file/1546310446/issues/source?bid=2038971&fileBranchId=3014792#l210)

File: weblate/trans/checks/format.py [hunk 7]
```diff
@@ -203,6 +203,13 @@ def check_single(self, source, target, unit):
...
+            ret.append((m.start(),m.group()))
```
Comment (nijel) [inline weblate/trans/checks/format.py]: Codacy Issue found: [Exactly one space required after comma](https://www.codacy.com/app/nijel/weblate/file/1546310446/issues/source?bid=2038971&fileBranchId=3014792#l211)

File: weblate/trans/checks/format.py [hunk 8]
```diff
@@ -203,6 +203,13 @@ def check_single(self, source, target, unit):
         '''
         return False
 
+    def check_highlight(self,source,unit):
```
Comment (nijel) [inline weblate/trans/checks/format.py]: Codacy Issue found: [Exactly one space required after comma](https://www.codacy.com/app/nijel/weblate/file/1546310446/issues/source?bid=2038971&fileBranchId=3014792#l206)

File: weblate/trans/checks/format.py [hunk 9]
```diff
@@ -203,6 +203,13 @@ def check_single(self, source, target, unit):
...
+        ret=[]
```
Comment (nijel) [inline weblate/trans/checks/format.py]: Codacy Issue found: [Exactly one space required around assignment](https://www.codacy.com/app/nijel/weblate/file/1546310446/issues/source?bid=2038971&fileBranchId=3014792#l207)

File: weblate/trans/checks/base.py [hunk 10]
```diff
@@ -133,6 +133,13 @@ def get_doc_url(self):
         return weblate.get_doc_url('user/checks', self.doc_id)
 
 
+    def check_highlight(self,source,unit):
```
Comment (nijel) [inline weblate/trans/checks/base.py]: Codacy Issue found: [Exactly one space required after comma](https://www.codacy.com/app/nijel/weblate/file/1549772583/issues/source?bid=2038971&fileBranchId=3014792#l136)

File: weblate/trans/checks/markup.py (hunk 1 above)
Comment (nijel) [inline weblate/trans/checks/markup.py]: Codacy Issue found: [More than one statement on a single line](https://www.codacy.com/app/nijel/weblate/file/1549772588/issues/source?bid=2038971&fileBranchId=3014792#l71)

Comment (nijel) [inline weblate/trans/checks/markup.py]: Codacy Issue found: [Exactly one space required around comparison](https://www.codacy.com/app/nijel/weblate/file/1549772588/issues/source?bid=2038971&fileBranchId=3014792#l71)

File: weblate/trans/checks/markup.py (hunk 2 above)
Comment (nijel) [inline weblate/trans/checks/markup.py]: Codacy Issue found: [Exactly one space required after comma](https://www.codacy.com/app/nijel/weblate/file/1549772588/issues/source?bid=2038971&fileBranchId=3014792#l72)

File: weblate/trans/checks/markup.py (hunk 3 above)
Comment (nijel) [inline weblate/trans/checks/markup.py]: Codacy Issue found: [Exactly one space required around assignment](https://www.codacy.com/app/nijel/weblate/file/1549772588/issues/source?bid=2038971&fileBranchId=3014792#l68)

File: weblate/trans/checks/markup.py (hunk 4 above)
Comment (nijel) [inline weblate/trans/checks/markup.py]: Codacy Issue found: [Exactly one space required after comma](https://www.codacy.com/app/nijel/weblate/file/1549772588/issues/source?bid=2038971&fileBranchId=3014792#l67)

File: weblate/trans/checks/markup.py (hunk 5 above)
Comment (nijel) [inline weblate/trans/checks/markup.py]: Codacy Issue found: [Final newline missing](https://www.codacy.com/app/nijel/weblate/file/1549772588/issues/source?bid=2038971&fileBranchId=3014792#l124)

File: weblate/trans/checks/format.py (hunk 6 above)
Comment (nijel) [inline weblate/trans/checks/format.py]: Codacy Issue found: [Exactly one space required around comparison](https://www.codacy.com/app/nijel/weblate/file/1549772587/issues/source?bid=2038971&fileBranchId=3014792#l210)

File: weblate/trans/checks/format.py (hunk 7 above)
Comment (nijel) [inline weblate/trans/checks/format.py]: Codacy Issue found: [Exactly one space required after comma](https://www.codacy.com/app/nijel/weblate/file/1549772587/issues/source?bid=2038971&fileBranchId=3014792#l211)

File: weblate/trans/checks/format.py (hunk 8 above)
Comment (nijel) [inline weblate/trans/checks/format.py]: Codacy Issue found: [Exactly one space required after comma](https://www.codacy.com/app/nijel/weblate/file/1549772587/issues/source?bid=2038971&fileBranchId=3014792#l206)

**Comment by lln-ijinus** [comment]
Just a screenshot to show what it does :
Alt+(char) option (or click):
image

Esc then number to dismiss check :
image

Alt then number to copy a machine translation :
image

**Comment by nijel** [comment]
Great idea, but could use some polishing (see test failures or automated comments above)...

File: weblate/trans/templatetags/translations.py [hunk 11]
```diff
@@ -114,6 +138,16 @@ def format_translation(value, language, diff=None, search_match=None,
             diffvalue = escape(force_unicode(diff[idx]))
             value = html_diff(diffvalue, value)
 
+        # Create span for checks highlights 
+        if highlights:
+            n=0
+            for (hidx,htext) in highlights:
+                p = value.find(htext,n)
+                if p >= 0:
+                    newpart=u'<span class="hlcheck">{0}</span>'.format(htext)
```
Comment (nijel) [inline weblate/trans/templatetags/translations.py]: Codacy Issue found: [Exactly one space required around assignment](https://www.codacy.com/app/nijel/weblate/file/1581742083/issues/source?bid=2038971&fileBranchId=3014792#l147)

File: weblate/trans/templatetags/translations.py [hunk 12]
```diff
@@ -114,6 +138,16 @@ def format_translation(value, language, diff=None, search_match=None,
...
+                    value=value[:p] + newpart +value[(p + len(htext)):]
```
Comment (nijel) [inline weblate/trans/templatetags/translations.py]: Codacy Issue found: [Exactly one space required around assignment](https://www.codacy.com/app/nijel/weblate/file/1581742083/issues/source?bid=2038971&fileBranchId=3014792#l148)

File: weblate/trans/templatetags/translations.py [hunk 13]
```diff
@@ -114,6 +138,16 @@ def format_translation(value, language, diff=None, search_match=None,
             diffvalue = escape(force_unicode(diff[idx]))
             value = html_diff(diffvalue, value)
 
+        # Create span for checks highlights
```
Comment (nijel) [inline weblate/trans/templatetags/translations.py]: Codacy Issue found: [Trailing whitespace](https://www.codacy.com/app/nijel/weblate/file/1581742083/issues/source?bid=2038971&fileBranchId=3014792#l141)

File: weblate/trans/templatetags/translations.py [hunk 14]
```diff
@@ -105,6 +105,30 @@ def format_translation(value, language, diff=None, search_match=None,
...
+                for n2 in xrange(n+1, len(highlights)):
+                    if n2 >= len(highlights):
+                        break
+                    eltest = highlights[n2]
+                    if eltest[0] >= elref[0] and eltest[0] < (elref[0] + len(elref[1])):
+                        highlights.pop(n2)
+                    elif eltest[0] > (elref[0] + len(elref[1])):
+                        break
+            #then transform highlights to escaped html
+            highlights = [ (h[0], escape(force_unicode(h[1]))) for h in highlights]
```
Comment (nijel) [inline weblate/trans/templatetags/translations.py]: Codacy Issue found: [Trailing whitespace](https://www.codacy.com/app/nijel/weblate/file/1581742083/issues/source?bid=2038971&fileBranchId=3014792#l130)

File: weblate/trans/templatetags/translations.py [hunk 15]
```diff
@@ -105,6 +105,30 @@ def format_translation(value, language, diff=None, search_match=None,
     parts = []
 
     for idx, value in enumerate(plurals):
+        highlights = None
+        # Find all checks highlight
+        if checks:
+            highlights = []
+            for c in checks:
+                highlights += c.check_highlight(value,None)
```
Comment (nijel) [inline weblate/trans/templatetags/translations.py]: Codacy Issue found: [Exactly one space required around assignment](https://www.codacy.com/app/nijel/weblate/file/1581742083/issues/source?bid=2038971&fileBranchId=3014792#l113)

File: weblate/trans/templatetags/translations.py [hunk 16]
```diff
@@ -114,6 +138,16 @@ def format_translation(value, language, diff=None, search_match=None,
             diffvalue = escape(force_unicode(diff[idx]))
             value = html_diff(diffvalue, value)
 
+        # Create span for checks highlights 
+        if highlights:
+            n=0
```
Comment (nijel) [inline weblate/trans/templatetags/translations.py]: Codacy Issue found: [Exactly one space required around assignment](https://www.codacy.com/app/nijel/weblate/file/1581742083/issues/source?bid=2038971&fileBranchId=3014792#l143)

File: weblate/trans/templatetags/translations.py [hunk 17]
```diff
@@ -105,6 +105,30 @@ def format_translation(value, language, diff=None, search_match=None,
...
+        highlights = None
+        # Find all checks highlight
+        if checks:
+            highlights = []
+            for c in checks:
+                highlights += c.check_highlight(value,None)
+            #Sort by order in string
+            if highlights: highlights.sort(key = lambda tup: tup[0])
+            #remove probelmatics ones (overlapping or one inside another)
+            for n in xrange(0,len(highlights)):
```
Comment (nijel) [inline weblate/trans/templatetags/translations.py]: Codacy Issue found: [Exactly one space required after comma](https://www.codacy.com/app/nijel/weblate/file/1581742083/issues/source?bid=2038971&fileBranchId=3014792#l117)

File: weblate/trans/templatetags/translations.py [hunk 18]
```diff
@@ -105,6 +105,30 @@ def format_translation(value, language, diff=None, search_match=None,
...
+                    if n2 >= len(highlights):
+                        break
+                    eltest = highlights[n2]
+                    if eltest[0] >= elref[0] and eltest[0] < (elref[0] + len(elref[1])):
+                        highlights.pop(n2)
+                    elif eltest[0] > (elref[0] + len(elref[1])):
+                        break
+            #then transform highlights to escaped html
+            highlights = [ (h[0], escape(force_unicode(h[1]))) for h in highlights]        
+
```
Comment (nijel) [inline weblate/trans/templatetags/translations.py]: Codacy Issue found: [Trailing whitespace](https://www.codacy.com/app/nijel/weblate/file/1581742083/issues/source?bid=2038971&fileBranchId=3014792#l131)

File: weblate/trans/templatetags/translations.py [hunk 19]
```diff
@@ -114,6 +138,16 @@ def format_translation(value, language, diff=None, search_match=None,
             diffvalue = escape(force_unicode(diff[idx]))
             value = html_diff(diffvalue, value)
 
+        # Create span for checks highlights 
+        if highlights:
+            n=0
+            for (hidx,htext) in highlights:
+                p = value.find(htext,n)
```
Comment (nijel) [inline weblate/trans/templatetags/translations.py]: Codacy Issue found: [Exactly one space required after comma](https://www.codacy.com/app/nijel/weblate/file/1581742083/issues/source?bid=2038971&fileBranchId=3014792#l145)

File: weblate/trans/templatetags/translations.py [hunk 20]
```diff
@@ -105,6 +105,30 @@ def format_translation(value, language, diff=None, search_match=None,
...
 
     for idx, value in enumerate(plurals):
+        highlights = None
+        # Find all checks highlight
+        if checks:
+            highlights = []
+            for c in checks:
+                highlights += c.check_highlight(value,None)
+            #Sort by order in string
+            if highlights: highlights.sort(key = lambda tup: tup[0])
```
Comment (nijel) [inline weblate/trans/templatetags/translations.py]: Codacy Issue found: [More than one statement on a single line](https://www.codacy.com/app/nijel/weblate/file/1581742083/issues/source?bid=2038971&fileBranchId=3014792#l115)

**Comment by lln-ijinus** [comment]
This time I installed pylint.
I let one error R0914(too-many-locals) because it need to do some code refactoring and I think you could prefer to do it your own way...
The modification also needs some new translations, but I don't know how you manage these...

**Comment by nijel** [comment]
Thanks! The only thing which I'm still unsure is using Alt+[letter] for the check shortcuts. Many of them are already used for other things and it might be confusing that some letters do manipulate with checks and others do not...

**Comment by lln-ijinus** [comment]
I agree that it's not really great to use Alt+Letter shortcuts but I didn't find anything else...
I tried to use functions keys but I had problem, same for Alt+Number that can't be overridden.

Another solution could be to use Alt+S then a number (and also Alt+M than a number for machine translation and Alt+C then a number for checks). Could be more consistent less prone to errors ?

Or if you have any other idea ?

**Comment by nijel** [comment]
Indeed having Alt+S/M/C then number sounds more consistent, this approach will also extent to other usages if needed. I'd like to hear feedback from others, maybe @scarabeusiv or @shanewims  (reporters of #894 and #953) can comment on this?

**Comment by shanewims** [comment]
I asked translators at vendor company for feedback on proposal: The proposal to introduce shortcuts is useful. It would be good to see an example of how the issues are highlighted when this is ready.

**Comment by nijel** [comment]
I thing latest proposal from @lln-ijinus sounds good enough. Will work on updating the code or should I look at it?

**Comment by nijel** [comment]
Okay, working on that.

**Pull request merged**